lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
   } else {
     session_map[imsi].push_back(std::move(session_state));
   }
-  MLOG(MINFO) << "Added a session (" << session_map[imsi].size()
-              << ") for IMSI " << imsi << " with session context ID ";
+  MLOG(MDEBUG) << "Added a session (" << session_map[imsi].size()
+               << ") for IMSI " << imsi << " with session context ID ";
   return true;
 }
 
-void SessionStateEnforcer::handle_session_init_rule_updates(
+bool SessionStateEnforcer::m5g_update_session_context(
     SessionMap& session_map, const std::string& imsi,
-    SessionState& session_state) {
-  uint32_t l_teid = 0;
-  auto itp        = pdr_map_.equal_range(imsi);
+    std::unique_ptr<SessionState>& session_state,
+    SessionUpdate& session_update) {
+  uint32_t gnode_teid = 0;
+  bool gnb_teid_get   = true;
+  bool upf_teid_get   = false;
+
+  /* Check and update latest session rules
+   * we get gnodeb TEID, and IP address details here
+   */
+  auto session_id                       = session_state->get_session_id();
+  SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];
+  gnode_teid                            = update_session_rules(
+      imsi, session_state, session_uc, gnb_teid_get, upf_teid_get);
+  if (!gnode_teid) return false;
+  uint32_t cur_version = session_state->get_current_version();
+  session_state->set_fsm_state(CREATED, &session_uc);
+  session_state->set_current_version(++cur_version, session_uc);
+  /* Reset the upf resend retransmission counter counter, send the session
+   * creation request to UPF
+   */
+  session_state->reset_rtx_counter();
+  m5g_send_session_request_to_upf(imsi, session_state);
+  return true;
+}
+
+uint32_t SessionStateEnforcer::update_session_rules(
+    const std::string& imsi, std::unique_ptr<SessionState>& session_state,
+    SessionStateUpdateCriteria& session_uc, bool gnb_teid_get,
+    bool upf_teid_get) {
+  uint32_t upf_teid = 0;
+  SetGroupPDR rule;
+
+  // Get the latest config
+  const auto& config = session_state->get_config();
+  auto itp           = pdr_map_.equal_range(imsi);
+  // Lets take local_teid of the session if already exists.
+  // Not needed in 2nd AMF request processing though
+  if (!upf_teid_get) {
+    upf_teid = session_state->get_upf_local_teid();
+  }
   for (auto itr = itp.first; itr != itp.second; itr++) {
     // Get the PDR numbers, now  get the rules from global static rule list
-    SetGroupPDR rule;
     GlobalRuleList.get_rule(itr->second, &rule);
-    session_state.insert_pdr(&rule);
-    // Get new TEID
-    if (!l_teid) l_teid = pipelined_client_->get_current_teid();
-    rule.mutable_pdi()->set_local_f_teid(pipelined_client_->get_next_teid());
+    // Get the UE ip address
+    rule.mutable_pdi()->set_ue_ip_adr(

```

Comment:
Let's pull the `rule.mutable_*` logic to construct the object into a separate function to clean things up.

Reply:
Created separate function for same.